File: python/course-repo/solution/04-concurrency-and-rate-limits/src/webhooks/activities.py
import httpx
from temporalio import activity

from .shared import WebhookDelivery


@activity.defn
def deliver_webhook(req: WebhookDelivery) -> int:
    activity.logger.info(
        "Delivering webhook for event %s to %s", req.event_id, req.url
    )
    # Idempotency-Key from Module 02 - keeps retries from duplicating this event.
    headers = {"Idempotency-Key": f"webhook:{req.event_id}"}
    response = httpx.post(req.url, json=req.payload, headers=headers, timeout=10.0)
    response.raise_for_status()
    return response.status_code

File: typescript/course-repo/exercise/02-idempotency-and-crash-safety/src/activities.ts
import { ApplicationFailure, activityInfo, log } from '@temporalio/activity';
import type { WebhookDelivery } from './shared';

export async function deliverWebhook(req: WebhookDelivery): Promise<number> {
  const { attempt } = activityInfo();
  log.info('Delivering webhook for event', { eventId: req.eventId, attempt });

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    // TODO: add an Idempotency-Key header derived from req.eventId.
    // The key must be deterministic across retries - use the logical webhook
    // event id (stable), not crypto.randomUUID() (different every attempt).
    // Format: 'Idempotency-Key': `webhook:${req.eventId}`
  };

  const response = await fetch(req.url, {
    method: 'POST',
    headers,
    body: JSON.stringify(req.payload),
  });
  if (!response.ok) throw new Error(`HTTP ${response.status}`);

  // Simulate a transient failure after the POST already landed.
  // Real-world equivalent: 500 from the receiver, network drop after the
  // receiver processed it, or the process crashed right after fetch resolved.
  // Temporal retries; without an idempotency key the receiver records every attempt.
  if (attempt < 3) {
    throw ApplicationFailure.create({
      message: `Simulated transient failure on attempt ${attempt}`,
    });
  }

  return response.status;
}
